hpcflow/sdk/helper/helper.py

```diff
@@ -255,14 +319,50 @@ def run_helper(
         timeout_check_interval = timedelta(seconds=timeout_check_interval_s)
 
     start_time = datetime.now()
+    end_time = start_time + timeout
     logger = get_helper_logger(app)
     controller = MonitorController(get_watcher_file_path(app), watch_interval, logger)
-    timeout_limit = timeout - timeout_check_interval
+    PID_vars = strip_helper_PID(app)
     try:
         while True:
-            if datetime.now() - start_time >= timeout_limit:
+            time_left_s = (end_time - datetime.now()).total_seconds()
+            if time_left_s <= 0:
                 helper_timeout(app, timeout, controller, logger)
-            time.sleep(timeout_check_interval_s)
+            time.sleep(min(timeout_check_interval_s, time_left_s))
+            # Reading args from PID file
+            PID_vars_new = strip_helper_PID(app)
+            for i in [1, 2, 3]:
+                if PID_vars_new[i] != PID_vars[i]:
+                    change = f"parameter from {PID_vars[i]} to {PID_vars_new[i]}."
+                    PID_vars[i] = PID_vars_new[i]
+                    if i == 1:
+                        timeout = float(PID_vars_new[i])
+                        if isinstance(timeout, timedelta):
+                            timeout_s = timeout.total_seconds()
+                        else:
+                            timeout_s = timeout
+                            timeout = timedelta(seconds=timeout_s)
+                        end_time = start_time + timeout
+                        logger.info(f"Updataed timeout {change}")
+                    elif i == 2:
+                        timeout_check_interval = float(PID_vars_new[i])
+                        if isinstance(timeout_check_interval, timedelta):
+                            timeout_check_interval_s = (
+                                timeout_check_interval.total_seconds()
+                            )
+                        else:
+                            timeout_check_interval_s = timeout_check_interval
+                            timeout_check_interval = timedelta(
+                                seconds=timeout_check_interval_s
+                            )
+                        logger.info(f"Updataed timeout_check_interval {change}")
+                    elif i == 3:
+                        watch_interval = float(PID_vars_new[i])
+                        controller.stop()
```

Comment:
Could you just add a `# TODO` here to note that we might later need to consider if a workflow change could be missed?

Reply:
Resolved in [e19c5ed](https://github.com/hpcflow/hpcflow-new/pull/340/commits/e19c5edf33d6194fcb135e9f7c3f0098cc445ac9)